Add Cloudflare APO cache purging for cache lifespan

Miraeld opened the pull request:
# Description

Fixes #6170 
- Add rocket_after_automatic_cache_purge hook listener to purge Cloudflare APO cache when cache lifespan expires
- Add after_rocket_clean_home hook listener to ensure home page cache is purged from Cloudflare APO
- Add purge_cloudflare_after_automatic_purge method that performs full cache purge for lifespan events
- Add comprehensive unit tests for new functionality
- Fixes issue #6170 where cache lifespan and rocket_clean_home() don't trigger Cloudflare APO cache purge


## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## Detailed scenario

### What was tested

Integration tests

### How to test

Refer to issue

## Technical description

### Documentation

This pull request adds support for purging the Cloudflare cache after an automatic cache purge (such as when the cache lifespan is reached), ensuring that Cloudflare's cache remains in sync with site changes. It also introduces unit tests to verify this new behavior and updates the list of subscribed events to include the new purge action.

**Cloudflare Cache Purge Enhancements**
* Added a new method `purge_cloudflare_after_automatic_purge` to the `Cloudflare` class, which triggers a full cache purge via Cloudflare when an automatic cache purge occurs.
* Registered the new event `rocket_after_automatic_cache_purge` in `get_subscribed_events` to call the new purge method, ensuring the integration is triggered at the correct time.

**Unit Test Additions**
* Added a unit test `purgeCloudflareAfterAutomaticPurge.php` to verify that the Cloudflare cache is purged when the plugin is active and not purged when inactive.
* Updated or added a unit test `getSubscribedEvents.php` to ensure the new event is correctly registered in the list of subscribed events.

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

Miraeld committed:
Add Cloudflare APO cache purging for cache lifespan

- Add rocket_after_automatic_cache_purge hook listener to purge Cloudflare APO cache when cache lifespan expires
- Add after_rocket_clean_home hook listener to ensure home page cache is purged from Cloudflare APO
- Add purge_cloudflare_after_automatic_purge method that performs full cache purge for lifespan events
- Add comprehensive unit tests for new functionality
- Fixes issue #6170 where cache lifespan and rocket_clean_home() don't trigger Cloudflare APO cache purge

Miraeld committed:
Fix ArgumentCountError in purge_cloudflare_after_automatic_purge

- Add 2 argument specification to hook registration
- Hook callback now correctly receives both $deleted and $args parameters
- Update corresponding unit test expectations

This fixes CI failures where the method expected 2 arguments but WordPress
was only passing 1 due to the default hook registration behavior.

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -72,6 +72,8 @@ public static function get_subscribed_events() {
 			'pre_get_rocket_option_do_cloudflare' => 'disable_cloudflare_option',
 			'rocket_after_clean_domain'           => 'purge_cloudflare',
 			'after_rocket_clean_files'            => 'purge_cloudflare_partial',
+			'after_rocket_clean_home'             => 'purge_cloudflare',
```
Do we really want to purge everything when the home cache is cleared @DahmaniAdame ?

CC @wordpressfan

DahmaniAdame replied:
This one should clear just the homepage.

Miraeld replied:
Fixed

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -358,6 +360,25 @@ public function purge_cloudflare_partial( $urls ) {
+		}
+
+		// For cache lifespan purge, we do a full cache purge since it affects the entire site.
+		$this->facade->purge_everything();
```
Should we be more precise and purge urls that were deleted, I believe we can get this from the deleted param in the hook. Maybe this is a product decision. @DahmaniAdame What do you think?

CC @wordpressfan

DahmaniAdame replied:
This one should clear specific URLs.

Miraeld replied:
Fixed

Miraeld committed:
Fix Cloudflare cache purging to be more precise

- Modify home cache purging to target only homepage instead of everything
- Implement cache file path to URL conversion for automatic cache purge
- Add support for specific URL purging instead of purging everything
- Update PHPStan baseline for new get_option() usage

Addresses feedback from PR review to make cache purging more targeted.

Miraeld committed:
Add comprehensive tests for Cloudflare cache purging improvements

- Create new test file for purge_cloudflare_home method with 5 test scenarios
- Update getSubscribedEvents test to reflect new home purge event mapping
- Completely rewrite purgeCloudflareAfterAutomaticPurge tests for new behavior
- Add coverage for plugin inactive, specific URL purging, and fallback scenarios

All tests validate the new precise cache purging behavior.

Miraeld committed:
Fix @since version tags to 3.20.2

Correct the version number in PHPDoc blocks for the new Cloudflare
cache purging methods to match the actual release version.

Miraeld committed:
Refactor: remove unnecessary else clause in purge method

Replace else clause with early return pattern to improve code clarity
while maintaining exact same functionality. Addresses Codacy suggestion
without breaking WordPress hook parameter compatibility.

Miraeld committed:
Fix Cloudflare cache purging to target specific URLs instead of everything

- Add targeted homepage purging for rocket_after_clean_home hook
- Add specific URL extraction for rocket_after_automatic_cache_purge hook
- Extract URLs from deleted cache file paths for precise Cloudflare purging
- Add comprehensive test coverage for new purging methods
- Maintain fallback to purge everything when specific targeting fails

Fixes feedback from PR review:
- Homepage cache clearing now purges only homepage instead of everything
- Automatic cache purge now targets specific deleted URLs instead of blanket clearing

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -341,6 +343,47 @@ public function purge_cloudflare() {
+
+		// If we can't determine specific post ID, fall back to purging everything.
+		// This ensures we don't miss anything important.
+		$this->facade->purge_everything();
```
Based on the thread [here](https://group-onecom.slack.com/archives/C08EFCYRQ2X/p1763462133128759?thread_ts=1763367090.157029&cid=C08EFCYRQ2X). We don't need a fallback

Miraeld replied:
Done

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -358,6 +401,42 @@ public function purge_cloudflare_partial( $urls ) {
+
+		if ( empty( $urls_to_purge ) ) {
+			// If no specific URLs found, fall back to purging everything as safety measure.
+			$this->facade->purge_everything();
```
Based on the thread [here](https://group-onecom.slack.com/archives/C08EFCYRQ2X/p1763462133128759?thread_ts=1763367090.157029&cid=C08EFCYRQ2X). We don't need a fallback

Miraeld replied:
Done

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -358,6 +401,42 @@ public function purge_cloudflare_partial( $urls ) {
+		}
+
+		// If no valid post IDs found, fall back to purging everything.
+		$this->facade->purge_everything();
```
Based on the thread [here](https://group-onecom.slack.com/archives/C08EFCYRQ2X/p1763462133128759?thread_ts=1763367090.157029&cid=C08EFCYRQ2X). We don't need a fallback

Miraeld replied:
Done

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -437,6 +516,88 @@ private function is_apo_enabled(): bool {
+					$urls[] = $url;
+				}
+			}
+		}
```
What do you think about a similar logic [here](https://github.com/wp-media/wp-rocket/blob/cd88a88ad917c73a2b889d76e41f153cea462cfc/inc/Engine/Preload/Subscriber.php#L401-L421) already being used in preload to extract the urls. I think it's less complex and has proven accurate over the years, maybe we can isolate the logic and re-use it on both sides.

Miraeld replied:
Done

Miraeld committed:
Address PR review feedback

- Remove fallback to purge_everything() in purge_cloudflare_home and purge_cloudflare_after_automatic_purge methods
- Simplify URL extraction logic using approach from Preload subscriber (proven accurate over years)
- Remove complex extract_urls_from_deleted_cache and cache_file_path_to_url helper methods
- Remove unused parameters ($root and $args) from method signatures
- Update hook subscriptions to reflect simpler method signatures
- Fix indentation in test files (tabs to spaces for consistency)
- Update unit tests to remove fallback scenarios and adjust for new logic
- All tests passing (9 assertions across 3 test files)

Miraeld committed:
Update PHPStan baseline for additional get_option usage

The new URL extraction logic uses get_option('permalink_structure') which adds one more usage to the baseline count (7 -> 8).

jeawhanlee commented on inc/ThirdParty/Plugins/CDN/Cloudflare.php:
```diff
@@ -358,6 +395,59 @@ public function purge_cloudflare_partial( $urls ) {
+					}
+				}
+
+				$urls_to_purge[] = str_replace( $data['home_path'], $data['home_url'], $file_path );
```
it would be nice to have this logic in a single place, shared with the preload Subscriber

jeawhanlee reviewed (approved):
Approving with a comment

jeawhanlee committed:
Abstract logic to process deleted url

jeawhanlee committed:
Used abstracted logic

jeawhanlee committed:
Updated name of method call

jeawhanlee committed:
Updated phpstan baseline

jeawhanlee committed:
Merge branch 'develop' into enhancement/6170-link-caches-tgt

jeawhanlee committed:
Ignored deprecation phpcs error

jeawhanlee committed:
2nd trial to ignore phpcs error

jeawhanlee committed:
Temporarily ignored phpcs deprecated error for PHP 8.5 on function finfo_open

jeawhanlee committed:
Reverted changes to ruleset

jeawhanlee committed:
Added inline comment for phpcs ignore

Mai-Saad commented:
@Miraeld @jeawhanlee Thanks for the PR, seems that we have a regression on the PR 
Related test case https://wpmediaqa.testrail.io/index.php?/tests/view/55345
steps:
1- website not using CF i.e served from nginx
2- set lifespan to 5 mins with filter
3- fresh install and activate of PR
4- wait till preload is done
5- wait 5 mins after any URL is completed in database
6- run "rocket_purge_time_event" ====> cached file from 5 mins isnot removed nor status change to pending in DB

hanna-meda commented:
Confirming it's the same for me. Using the same steps as above, I get the same outcome. Returning this to In Progress because of this.

jeawhanlee commented:
@Mai-Saad @hanna-meda I'm not able to reproduce this on my personal test site and also on newlabs.

hanna-meda commented:
@jeawhanlee Thanks for checking this out. Indeed I can no longer reproduce either. Tried again on newlabs and on rocketlabs. Could have been a timing mismatch (DB status vs files vs cron run time). 
Taking this back to QA and resume testing.

hanna-meda reviewed (approved):
QA Done. This can be merged. 

Regarding the failed tests from TR:
[testrail-report-794.pdf](https://github.com/user-attachments/files/23824208/testrail-report-794.pdf)
Two new issues were logged to be handled later:
https://github.com/wp-media/wp-rocket/issues/7922
https://github.com/wp-media/wp-rocket/issues/7916

wordpressfan committed:
Merge branch 'develop' into enhancement/6170-link-caches-tgt
